## 1-qadam. Kutubxonalarni import qilish

In [25]:
import numpy as np # chiziqli algebra
import pandas as pd # ma'lumotlarga ishlov berish, CSV fayliga yozish/o’qish
import matplotlib.pyplot as plt # ma'lumotlarni vizualizatsiya qilish uchun
import seaborn as sns # statistik ma'lumotlarni vizualizatsiya qilish uchun
%matplotlib inline

## 2-qadam. Ma’lumotlar to’plamini import qilish

In [2]:
data = 'data2.csv'
df = pd.read_csv(data, sep=',')

## 3-qadam. Tadqiqot ma’lumotlarini tahlil qilish

#### 3.1. Ma’lumotlar to’plamlarini o’lchamini ko’rish

In [3]:
df.shape

(104, 7)

#### 3.2. Ma’lumotlar to’plamining dastlabki besh qatorini ko’rsatish

In [4]:
df.head(10)

qon_bosimi_speed  qon_bosimi_count  xolesterin  bmi  turmush_tarzi  \
0               130                85         100   18              1   
1               140                99         130   40              0   
2               130                85         100   18              1   
3               140                99         130   40              0   
4               135                82         100   19              1   
5               142                97         130   35              0   
6               144                99         130   37              0   
7               132                88         100   18              1   
8               141                95         130   34              0   
9               132                85         100   20              1   

   tibbiyot_tarixi  dastlabki_holat  
0                1                1  
1                1                0  
2                1                1  
3                1                0  
4                1                1  
5                0                0  
6                1                0  
7                0                1  
8                0                0  
9                1                1

#### 3.3. Ustun nomlarini qayta nomlash

In [5]:
col_names = ['qon_bosimi_speed','qon_bosimi_count','xolesterin','bmi','turmush_tarzi','tibbiyot_tarixi','dastlabki_holat']
col_names

['qon_bosimi_speed',
 'qon_bosimi_count',
 'xolesterin',
 'bmi',
 'turmush_tarzi',
 'tibbiyot_tarixi',
 'dastlabki_holat']

#### 3.4. Ma’lumotlar to’plamining xulosasini ko’rish

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 104 entries, 0 to 103
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   qon_bosimi_speed  104 non-null    int64
 1   qon_bosimi_count  104 non-null    int64
 2   xolesterin        104 non-null    int64
 3   bmi               104 non-null    int64
 4   turmush_tarzi     104 non-null    int64
 5   tibbiyot_tarixi   104 non-null    int64
 6   dastlabki_holat   104 non-null    int64
dtypes: int64(7)
memory usage: 5.8 KB


#### 3.5. Ma’lumotlar to’plamini kategorik va raqamli o’zgaruvchilarga ajratish

##### 3.5.1. Kategorik o’zgaruvchilarni aniqlash

In [7]:
categorical = [var for var in df.columns if df[var].dtype=='int64']
print('Bu yerda {} ta kategorik o\'zgaruvchi bor\n'.format(len(categorical)))
print('Kategorik o’zgaruvchilar :\n\n', categorical)

Bu yerda 7 ta kategorik o'zgaruvchi bor

Kategorik o’zgaruvchilar :

 ['qon_bosimi_speed', 'qon_bosimi_count', 'xolesterin', 'bmi', 'turmush_tarzi', 'tibbiyot_tarixi', 'dastlabki_holat']


In [8]:
categorical = [var for var in df.columns if df[var].dtype=='O']
print('Bu yerda {} ta kategorik o\'zgaruvchi bor\n'.format(len(categorical)))
print('Kategorik o’zgaruvchilar :\n\n', categorical)

Bu yerda 0 ta kategorik o'zgaruvchi bor

Kategorik o’zgaruvchilar :

 []


In [9]:
# Kategorik o’zgaruvchilarni ko’rish
# df[categorical].head()

##### 3.5.2. Kategorik o'zgaruvchilarda yetishmayotgan qiymatlarni ko’rish

In [10]:
df[col_names].isnull().sum()

qon_bosimi_speed    0
qon_bosimi_count    0
xolesterin          0
bmi                 0
turmush_tarzi       0
tibbiyot_tarixi     0
dastlabki_holat     0
dtype: int64

##### 3.5.4. Kategorik o'zgaruvchilarning chastota taqsimotini ko'rish

In [11]:
 for var in col_names: 
    print(df[var].value_counts()/float(len(df)))

qon_bosimi_speed
132    0.182692
140    0.115385
130    0.086538
142    0.076923
182    0.057692
135    0.048077
133    0.048077
122    0.038462
128    0.038462
134    0.028846
129    0.028846
152    0.028846
136    0.028846
131    0.028846
145    0.028846
141    0.028846
144    0.028846
153    0.019231
123    0.019231
125    0.019231
194    0.009615
187    0.009615
Name: count, dtype: float64
qon_bosimi_count
86     0.201923
106    0.105769
85     0.096154
99     0.096154
97     0.076923
88     0.057692
95     0.057692
82     0.038462
105    0.028846
76     0.028846
90     0.028846
87     0.028846
96     0.028846
108    0.019231
84     0.019231
83     0.019231
156    0.019231
102    0.019231
109    0.019231
94     0.009615
Name: count, dtype: float64
xolesterin
100    0.250000
130    0.230769
110    0.086538
150    0.048077
107    0.048077
120    0.048077
111    0.048077
140    0.038462
210    0.038462
105    0.028846
108    0.028846
115    0.019231
103    0.019231
122    0.019231
180

#### 3.6. Kardinallikni aniqlash

In [12]:
for var in col_names:
        print(var, ' ustuni tarkibida ', len(df[var].unique()), ' xil nishon bor')

qon_bosimi_speed  ustuni tarkibida  22  xil nishon bor
qon_bosimi_count  ustuni tarkibida  20  xil nishon bor
xolesterin  ustuni tarkibida  18  xil nishon bor
bmi  ustuni tarkibida  22  xil nishon bor
turmush_tarzi  ustuni tarkibida  2  xil nishon bor
tibbiyot_tarixi  ustuni tarkibida  2  xil nishon bor
dastlabki_holat  ustuni tarkibida  2  xil nishon bor


'qon_bosimi_speed' va 'bmi' ustunlari tarkibida eng ko’p nishonlar borligi aniqlandi

#### 3.5.1-3.5.4 qadamlar raqamli o’zgaruvchilar uchun aniqlanadi

In [13]:
numerical = [var for var in df.columns if df[var].dtype!='O']
print('Bu yerda{} ta raqamli o’zgaruvchi bor \n'.format(len(numerical)))
print('Raqamli o’zgaruvchilar:', numerical)

Bu yerda7 ta raqamli o’zgaruvchi bor 

Raqamli o’zgaruvchilar: ['qon_bosimi_speed', 'qon_bosimi_count', 'xolesterin', 'bmi', 'turmush_tarzi', 'tibbiyot_tarixi', 'dastlabki_holat']


## 4-qadam. Belgilar vektori va maqsadli o'zgaruvchini e'lon qilish

In [14]:
X = df.drop(['dastlabki_holat'], axis=1)
y = df['dastlabki_holat']

### 5-qadam. Ma'lumotlarni alohida o’quv va test to'plamiga bo'lish

In [15]:
# X va y ni o'quv va test to'plamlariga ajratish

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 0)
X_train.shape, X_test.shape

((72, 6), (32, 6))

### 6-qadam. Belgilarni ishlab chiqish

#### 6.1. Ishga tayyorlash

In [16]:
# kategorik o'zgaruvchilarni ko'rsatish
categorical = [col for col in X_train.columns if X_train[col].dtypes == 'O']
categorical

# raqamli o'zgaruvchilarni ko'rsatish
numerical = [col for col in X_train.columns if X_train[col].dtypes != 'O']
numerical

# o'quv to'plamidagi kategorik o'zgaruvchilarda yetishmayotgan qiymatlarning foizini chop etish
X_train[categorical].isnull().mean()

# kategoriyali o'zgaruvchilarni yetishmayotgan ma'lumotlar bilan chop etish
for col in categorical:
    if X_train[col].isnull().mean()>0:
        print(col, (X_train[col].isnull().mean()))
        
# eng tez-tez uchraydigan qiymatga ega bo'lmagan kategorik o'zgaruvchilarni hisoblash
# for df2 in [X_train, X_test]:
    # df2
    # df2['qon_bosimi_speed'].fillna(X_train['qon_bosimi_speed'].mode()[0], inplace=True)
    # df2['Kasb'].fillna(X_train['Kasb'].mode()[0], inplace=True)
    # df2['Mamlakat'].fillna(X_train['Mamlakat'].mode()[0], inplace=True)  


#### 6.2. Kategoriyali o’zgaruvchilarni kodlash (__Bajarilgan__)

In [17]:
# encoder = ce.OneHotEncoder(cols=['Ishchining_sinfi', 'Ma\'lumoti', 'Oilaviy_holati', 'Kasb', 'Munosabati', 'Irq', 'Jins', 'Mamlakat'])

# X_train = encoder.fit_transform(X_train)
# X_test = encoder.transform(X_test)

# X_train.head()


Model yaratish uchun o'quv va test majmuasi tayyor. Bundan oldin, barcha belgi o'zgaruvchilarni bir xil masshtabga solishimiz kerak. Bu __belgini masshtablash__ deb ataladi

### 7-qadam. Belgini masshtablash

In [18]:
cols = X_train.columns

from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
X_train = pd.DataFrame(X_train, columns=[cols])
X_test = pd.DataFrame(X_test, columns=[cols])

X_train.head()

qon_bosimi_speed qon_bosimi_count xolesterin       bmi turmush_tarzi  \
0        -0.044444        -0.145455  -0.033333  1.200000          -1.0   
1         1.022222         0.800000   0.633333  1.066667          -1.0   
2         0.844444         0.509091   0.633333  1.000000          -1.0   
3        -0.044444        -0.872727  -0.366667  0.066667           0.0   
4        -0.044444        -0.145455  -0.033333 -0.133333           0.0   

  tibbiyot_tarixi  
0             0.0  
1             0.0  
2             0.0  
3             0.0  
4             0.0

Shunday qilib, bu bosqichlardan so’ng X_train ma’lumotlar to’plami Gauss Naïve Bayes klassifikatoriga kirish uchun tayyor bo’ladi

### 8-qadam. Modelni o’qitish

In [27]:
# O’quv to'plamida Gauss Naive Bayes klassifikatorini o’qitish
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()
gnb.fit(X_train, y_train)

GaussianNB()

### 9-qadam. Natijani bashorat qilish

In [26]:
y_pred = gnb.predict(X_test)
y_pred

array([0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       0, 1, 0, 0, 0, 0, 1, 0, 1, 1], dtype=int64)

### 10-qadam. Modelning aniqligini tekshirish

In [28]:
from sklearn.metrics import accuracy_score
print('Modelning aniqligi bali: {0:0.4f}'.format(accuracy_score(y_test, y_pred)))

Modelning aniqligi bali: 1.0000


Bu yerda y_test haqiqiy sinf yorliqlari va y_pred test to'plamidagi taxmin qilingan sinf belgilaridir

#### 10.1. O’quv to'plami va test to'plamining aniqligini solishtirish

In [22]:
print('O’quv to’plami bali: {:.4f}'.format(gnb.score(X_train, y_train)))
print('Test to’plami bali: {:.4f}'.format(gnb.score(X_test, y_test)))

O’quv to’plami bali: 1.0000
Test to’plami bali: 1.0000


Xulosa: O’quv to'plamining aniqligi balli 1.0000, test to'plamining aniqligi ham 1.00000. Bu ikki qiymat bir xil. Shunday qilib, qayta o’qitishga muhtoj emasmiz.

#### 10.2. Test to'plamidagi sinf taqsimotini tekshirish

In [23]:
y_test.value_counts()

dastlabki_holat
0    16
1    16
Name: count, dtype: int64

In [24]:
null_accuracy = (16/(16+16))
print(' Null  aniqlik balli: {0:0.4f}'. format(null_accuracy))

 Null  aniqlik balli: 0.5000


Ko'rishimiz mumkinki, eng tez-tez uchraydigan sinfdagi hodisalar 16 ni tashkil qiladi. Shunday qilib, 16 ni umumiy hodisalar soniga bo'lish orqali nol aniqlikni hisoblashimiz mumkin.

Bizning model aniqlik ballimiz 0.5000 ekanligini ko'rishimiz mumkin. Shunday qilib, bizning Gauss Naive Bayes tasnifi modelimiz sinf belgilarini bashorat qilishda juda yaxshi ishlamoqda degan xulosaga kelishimiz mumkin.
Endi yuqoridagi tahlilga asoslanib, tasniflash modelimiz aniqligi juda yaxshi degan xulosaga kelishimiz mumkin. Modelimiz sinf belgilarini bashorat qilish nuqtai nazaridan juda yaxshi ishlamoqda.
